ee/tabby-db/src/github_oauth_credential.rs

```diff
@@ -31,20 +31,25 @@ impl DbConn {
     pub async fn update_github_oauth_credential(
         &self,
         client_id: &str,
-        client_secret: &str,
+        client_secret: Option<&str>,
         active: bool,
     ) -> Result<()> {
-        let client_id = client_id.to_string();
-        let client_secret = client_secret.to_string();
+        let mut sql = r#"INSERT INTO github_oauth_credential (id, client_id, client_secret)
+                                VALUES (:id, :cid, :secret) ON CONFLICT(id) DO UPDATE "#
+            .to_string();
+        if client_secret.is_some() {
+            sql += r#"SET client_id = :cid, client_secret = :secret, active = :active, updated_at = datetime('now')
+                    WHERE id = :id"#;
+        } else {
+            sql += r#"SET client_id = :cid, active = :active, updated_at = datetime('now')
+                    WHERE id = :id"#;
+        }
 
+        let client_id = client_id.to_string();
+        let client_secret = client_secret.unwrap_or_default().to_owned();
```

Comment:
shall we avoid unwrap but restructure the condition in a different way?

Reply:
what's different way, specifically?

Reply:
```rust
let (param, query) = if let Some(client_secret) = client_secret {
  // construct sql query and param object with client_secret and return.
} else {
  // construct sql query and param object without client_secret and return.
}

execute sql query and param...
```

Reply:
As I tried, return `param` from `if let ... else ...` can not compile, it gives error:
```
temporary value dropped while borrowed
```
the `named_params!` macro generates value with type `&[(&str, &dyn ToSql)]`, which will be freed at the end of `if else` statement.

Reply:
also, we need to do `INSERT INTO github_oauth_credential (id, client_id, client_secret)` for first write.
If sceret is not provided, we need the default empty string to be filled, as db schema requires client_secret is `NOT NULL`.

Reply:
> the named_params! macro generates value with type &[(&str, &dyn ToSql)], which will be freed at the end of if else statement.

in that case, let's just write everything down, and try extract common structures between each branch.

> If sceret is not provided, we need the default empty string to be filled, as db schema requires client_secret is NOT NULL.

but when client_secret is empty and client_secret is empty, we should just probably just explicit returns an error?

Reply:
> > the named_params! macro generates value with type &[(&str, &dyn ToSql)], which will be freed at the end of if else statement.
> 
> in that case, let's just write everything down, and try extract common structures between each branch.

I don't see much value here to extract `params` structure, either `client_secret` is some string or empty, we need these 4 parameters to execute the sql statement.

> > If sceret is not provided, we need the default empty string to be filled, as db schema requires client_secret is NOT NULL.
> 
> but when client_secret is empty and client_secret is empty, we should just probably just explicit returns an error?

Just like `register` and `tokenAuth` we can add input validation to return user-friendly error message for your concern, do you want to add it?